### ISSUE TITLE
Improve Curl Error Checking

### ISSUE THREAD

**Issue by bauzer714** [boredazfcuk/docker-icloudpd#127]
The curl requests only check if curl exited with an error when it should only print "successfully" if the HTTP Status was a 200 series response. 

I implemented a firewall rule that was too aggressive and was blocking the request. The logging of icloudPD indicated the webhook was successful, but it definitely wasn't.

Here are a few samples where 400 / 500 series responses IMO should fail, but they print success as the error code is 0.

![image](https://user-images.githubusercontent.com/13224936/152200987-64be9950-2dbc-4247-b6e3-48df9dd08d22.png)

**Comment by boredazfcuk** [comment]
Yeah, I did discover this myself a while back. It's because curl will exit successfully if it successfully posts data to the API that it hits... It doesn't matter if it posts the wrong info and the website on the other end rejects it, curl did it's job.

I did try curl with the `--fail` command line argument, but that still exits successfully with certain https responses.

The only way I could do this properly is to capture the output of the curl command and parse it for a 200 status code. Just means a bit of work for something that doesn't really affect me, and can be a bit of a pain to test. I'll leave this open and look into it at some point down the line.

**Comment by bauzer714** [comment]
I've done this for other personal projects I can copy code from. I'll do it for one of them as a sample and get your support before replicating to the entire project.

Sound good?

**Comment by boredazfcuk** [comment]
Cool. Are you using the new `--fail-with-body` argument added to curl 7.76?

The version of curl inside the container is 7.79, so we have the option to use that, if it simplifies things.

**Comment by boredazfcuk** [comment]
I actually had a quick look at this just after I finished work and have pushed a new version of the code. I had a look at parsing the curl output body, but it was in HTML and found that just parsing the API's JSON response was a little easier, so I've done it that way.

It would need to be customised for each notification type though, so it may mean we still need the curl code for other APIs.

**Comment by bauzer714** [comment]
From my experience, relying on HTTP status is more resilient.  You mention "the API's JSON response", which will work for some of the specific integrations utilized (like you did for Telegram). For Webhook, it could be any endpoints and responses. Status code would work better.  I'd recommend the ones where it's possible to utilize a known "Good" response, implement those - for Webhook, use status code.  

If you're good with that, I can submit a more formal pull request - or you can grab the code and implement it yourself.

**Comment by boredazfcuk** [comment]
I think I've about sorted this now. I looked at Prowl and that returns info in XML rather than JSON, so it's a pain parse the output.

Just went with curl error code checking across the board in the end. Keeps it simpler so there's less to go wrong... also meant I could de-duplicate code as well, by checking the curl result right at the end of the function instead of doing it per notification type.

I've also standardised the notification messages across the notification types and tidied up the code in places to remove duplication. I've pushed the new version to Dockerhub, so give it a go and let me know if you find any problems.

**Comment by boredazfcuk** [comment]
Oh and one other thing, I've implemented it so that it just checks for a 200 status code, everything else is a fail.

I did notice that you're doing some regex testing on the error codes in your gists... Do you actually see non 200 success codes returned from webhook? If you do, the code may need looking at again... alpine Linux in the container runs dash, which is posix compatible, so it throws and error when it encounters double square brackets - [[ ]]. I'm not 100% sure it can do the `=~` comparison either.

**Comment by bauzer714** [comment]
I think that's good enough, can revisit in the future if someone has a problem. Appreciate the help. I'll pull the latest container and test early next week.
